src/Swarm/Language/LSP/VarUsage.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Swarm.Language.LSP.VarUsage where
+
+import Control.Monad (guard)
+import Data.Map (Map)
+import Data.Map qualified as M
+import Data.Set (Set)
+import Data.Set qualified as S
+import Data.Text (Text)
+import Data.Text qualified as T
+import Language.LSP.Types qualified as J
+import Swarm.Language.Parse qualified as P
+import Swarm.Language.Syntax
+import Swarm.Util qualified as U
+
+data BindingType
+  = Lambda
+  | Let
+  | Bind
+  deriving (Eq, Show)
+
+data VarUsage = VarUsage LocVar BindingType
+
+type BindingSites = Map Var (BindingType, Syntax)
+
+data Usage = Usage
+  { usages :: Set Var
```

Comment:
It's not quite correct to just have a `Set Var` for `usages`, since variables are not uniquely determined by their name alone, but rather by their name + location of their binding site.  I think this should be a `Set LocVar` instead.

Concretely, if you try something like this:
```
def y = \x. \x. x + 1 end
```
no warning is generated, since the usage of `x` makes the `checkOccurrences` function think everything is OK for both binders. But in fact, there *should* be a warning, since the first `x` is not used.

Reply:
Got it.  Now handled.  Add unit tests, including for a shadowed variable that has usage between the first and second binding site.